test/common/http/filter/ext_authz_test.cc

```diff
@@ -0,0 +1,304 @@
+  EXPECT_CALL(filter_callbacks_, encodeHeaders_(HeaderMapEqualRef(&response_headers), true));
+  EXPECT_CALL(filter_callbacks_, continueDecoding()).Times(0);
+  EXPECT_CALL(filter_callbacks_.request_info_,
+              setResponseFlag(Envoy::RequestInfo::ResponseFlag::Unauthorized));
+  request_callbacks_->onComplete(Envoy::ExtAuthz::CheckStatus::Denied);
+
+  EXPECT_EQ(
+      1U,
+      cm_.thread_local_cluster_.cluster_.info_->stats_store_.counter("ext_authz.denied").value());
+  EXPECT_EQ(
+      1U,
+      cm_.thread_local_cluster_.cluster_.info_->stats_store_.counter("upstream_rq_4xx").value());
+  EXPECT_EQ(
+      1U,
+      cm_.thread_local_cluster_.cluster_.info_->stats_store_.counter("upstream_rq_403").value());
+}
+
+// Test that when a connection awaiting a authorization response is canceled then the authorization
+// call is closed.
+TEST_P(HttpExtAuthzFilterParamTest, ResetDuringCall) {
+  InSequence s;
+
+  ON_CALL(filter_callbacks_, connection()).WillByDefault(Return(&connection_));
+  EXPECT_CALL(connection_, remoteAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(connection_, localAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(*client_, check(_, _, _))
+      .WillOnce(WithArgs<0>(Invoke([&](Envoy::ExtAuthz::RequestCallbacks& callbacks) -> void {
+        request_callbacks_ = &callbacks;
+      })));
+
+  EXPECT_EQ(FilterHeadersStatus::StopIteration, filter_->decodeHeaders(request_headers_, false));
+
+  EXPECT_CALL(*client_, cancel());
+  filter_->onDestroy();
+}
+
+// Check a bad configuration results in validation exception.
+TEST_F(HttpExtAuthzFilterTest, BadConfig) {
+  const std::string filter_config = R"EOF(
+  failure_mode_allow: true
+  grpc_service: {}
+  )EOF";
+
+  envoy::config::filter::http::ext_authz::v2::ExtAuthz proto_config{};
+  MessageUtil::loadFromYaml(filter_config, proto_config);
+
+  EXPECT_THROW(
+      MessageUtil::downcastAndValidate<const envoy::config::filter::http::ext_authz::v2::ExtAuthz&>(
+          proto_config),
+      ProtoValidationException);
+}
+
+// Test when failure_mode_allow is NOT set and the response from the authorization service is Error
+// that the request is not allowed to continue.
+TEST_F(HttpExtAuthzFilterTest, ErrorFailClose) {
+  const std::string fail_close_config = R"EOF(
+  grpc_service:
+    envoy_grpc:
+      cluster_name: "ext_authz_server"
+  failure_mode_allow: false
+  )EOF";
+  initialize(fail_close_config);
+  InSequence s;
+
+  ON_CALL(filter_callbacks_, connection()).WillByDefault(Return(&connection_));
+  EXPECT_CALL(connection_, remoteAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(connection_, localAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(*client_, check(_, _, _))
+      .WillOnce(WithArgs<0>(Invoke([&](Envoy::ExtAuthz::RequestCallbacks& callbacks) -> void {
+        request_callbacks_ = &callbacks;
+      })));
+
+  EXPECT_EQ(FilterHeadersStatus::StopIteration, filter_->decodeHeaders(request_headers_, false));
```

Comment:
Should there be a similar validation in the NoCluster and NoRoute tests?  Those should have different behavior depending on the value of setting of failure_mode_allow, right?

Reply:
@ggreenway The NoCluster and NoRoute scenarios don't change behavior with `failure_mode_allow`. In both the cases the `state_` does not change to `State::Calling`.

Reply:
Looking at the documentation for failure_mode_allow, it says "The filter's behaviour in case the external authorization service does not respond back." Isn't that the case for both NoCluster and NoRoute? So shouldn't the behavior for both of these cases depend on the value for failure_mode_allow?  Or am I misunderstanding something?

Reply:
@ggreenway As i understand the NoRoute/NoCluster are about the http request's route and cluster not being available. In that case this filter would be a no-op.
The `failure_mode_allow` is concerned with behavior when the external gRPC service returns an error (typically when the authorization service does not respond back).
Or am I totally missing/messing up the logic here?

Reply:
At least in my mind, NoRoute/NoCluster seem like they should be treated the same way as the external grpc returning an error.  This is just the transport to talk to the external grpc returning an error.  In both cases, we failed to get a result on whether this request is authorized or not, just from different layers of the stack.  So it seems like both cases should fallback to however failure_mode_allow is configured. But I could be convinced otherwise; that's just how I would have interpreted the settings and documentation.

Reply:
It's not unreasonable to attribute to the normal selected cluster, since this is the service for which there was a denial. The denial though is really a function of everything that gets sent to the auth server, i.e. all attributes. I was just surprised that we're not consistent between TCP and HTTP filters.

Reply:
thanks for the clarification. let me address the specific feedback here. Once that is done I'll re-visit the TCP filter and check if I can be made more consistent.

Reply:
@ggreenway I think this is a bit different actually. NoRoute/NoCluster in this code are not errors from the external gRPC transport but properties of the data plane request being forwarded. There should likely always be a route in any expected configuration of the ext_authz filter.

Reply:
@saumoh I chatted with @ggreenway and we agree that the last patch you added is probably not needed. Can you back it out and then we can merge this? Thanks!

Reply:
@htuch @ggreenway kk. let me back it out then. Thanks.

Reply:
@saumoh Sorry for the extra work; I misunderstood what the check for NoRoute/NoCluster was doing. I thought it was on the auth connection/request, but it was actually for the original/main request.

Reply:
@ggreenway No worries. thanks for the reviews...it made the code much better and I learned quite a few new things with the feedback.